File: src/user-cpp/share/module.cpp
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

#include <string>
#include <iostream>

#include <skullcpp/api.h>
#include "skull_metrics.h"
#include "skull_protos.h"
#include "config.h"

// Create Metrics
static skullcpp::metrics::module moduleMetrics;

/**
 * Module Initialization Function.
 *
 * @note This function be called during the starting phase before the server
 *        is ready for serving
 *
 * @return
 *    - 0:     Initialization success
 *    - non-0: Initialization failure
 */
static int module_init(const skull_config_t* config) {
    SKULLCPP_LOG_INFO("ModuleInit", "module({MODULE_NAME}): init");

    // Load skull_config to skullcpp::Config
    auto& conf = skullcpp::Config::instance();
    conf.load(config);
    return 0;
}

/**
 * Module Release Function
 *
 * @note This function be called during the server shutdown phase
 */
static void module_release() {
    SKULLCPP_LOG_INFO("ModuleRelease", "module({MODULE_NAME}): released");
}

/**
 * Unpack the request from raw bytes stream.
 *
 * @note This function only be executed when this module is the first module
 *        in the workflow
 *
 * @return
 *   - = 0: Still need more data
 *   - > 0: Already unpacked a completed request, the returned value will be
 *          consumed from txn input buffer
 *   - < 0: Error occurred
 */
static ssize_t module_unpack(skullcpp::Txn& txn, const void* data,
                             size_t data_sz) {
    // Increase the 'module.request' metrics counter
    moduleMetrics.request.inc(1);

    SKULLCPP_LOG_INFO("ModuleUnpack",
                      "module_unpack({MODULE_NAME}): data sz:" << data_sz);

    // Deserialize data to transaction data
    auto& sharedData = (skull::workflow::{IDL_NAME}&)txn.data();
    sharedData.set_data(data, data_sz);
    return (ssize_t)data_sz;
}

/**
 * Module runnable function, when the transaction is moved to this module, this
 *  function will be executed.
 *
 * @return
 *  -     0: Everything is ok, transation will move to next module
 *  - non-0: Error occurred, transaction will be cancelled, and will run the
 *          `pack` function of the last module
 */
static int module_run(skullcpp::Txn& txn) {
    auto& sharedData = (skull::workflow::{IDL_NAME}&)txn.data();

    // Print and Log the current shared data value
    SKULLCPP_LOG_INFO("ModuleRun", "receive data: " << sharedData.data());
    return 0;
}

/**
 * Pack the data to `TxnData`, the data will be sent back to the sender. This
 *  function only be called when this module is the last module in the workflow
 */
static void module_pack(skullcpp::Txn& txn, skullcpp::TxnData& txndata) {
    auto& sharedData = (skull::workflow::{IDL_NAME}&)txn.data();

    // If the transaction status is not OK, return a 'error' message
    if (txn.status() != skullcpp::Txn::TXN_OK) {
        SKULLCPP_LOG_ERROR("ModulePack",
                           "module_pack({MODULE_NAME}): error status occurred: "
                               << txn.status()
                               << ". txn data: " << sharedData.data(),
                           "This error is expected");
        txndata.append("error");
        return;
    }

    // Increase the 'module.response' metrics counter
    moduleMetrics.response.inc(1);

    // Log and return response message
    SKULLCPP_LOG_INFO("ModulePack", "module_pack({MODULE_NAME}): data sz:"
                                        << sharedData.data().length());
    txndata.append(sharedData.data());
}

/******************************** Register Module *****************************/
static skullcpp::ModuleEntry module_entry = {
    module_init,
    module_release,
    module_run,
    module_unpack,
    module_pack
};

SKULLCPP_MODULE_REGISTER(&module_entry)

File: src/user-py/share/module.py
"""Skull Module"""

import pprint

from skull import logger
from skull.txn import Txn
from skull.txndata import TxnData

from common import metrics
from common.proto import *


def module_init(config):
    """Module Init Entry.

    It's called once when module be loaded

    @param config  A parsed yamlObj

    @return - True:  Initialization success
            - False: Initialization failure
    """

    logger.info('ModuleInit', 'config: {}'.format(pprint.pformat(config)))
    return True


def module_release():
    """Module Release Function, it's called once when module be unloaded"""

    logger.debug("py module release")


def module_unpack(txn: Txn, data: bytes):
    """Input data unpack function.

    It's called if this module is the 'first' module in the workflow and there
    is input data incoming

    @param txn  Transaction context which is used for getting shared
                transaction data or invoking `service.call()`
    @param data Input read-only `bytes` data

    @return - > 0: Tell engine N bytes be consumed
            - = 0: Need more data
            - < 0: Error occurred
    """

    logger.debug('ModuleUnpack: receive data: {}'.format(data.decode('UTF-8')))

    # Store data into txn sharedData
    sharedata = txn.data()
    sharedata.data = data
    return len(data)


def module_pack(txn: Txn, txndata: TxnData):
    """Output data pack function.

    It's called after 'module_run' function.
     the workflow (It would no effect if there is no response needed)

    @param txn  Transaction context which is used for getting shared
                transaction data or invoking `service.call()`
    @param data Input data

    @return How many bytes be consumed
    """

    logger.debug("py module pack")

    # Increase counters
    mod_metrics = metrics.module()
    mod_metrics.response.inc(1)

    # Assemble response
    if txn.status() != Txn.TXN_OK:
        txndata.append('error')
    else:
        sharedData = txn.data()
        logger.debug("pack data: %s" % sharedData.data)
        logger.info('ModulePack', 'module_pack: data sz: {}'.format(
            len(sharedData.data)))

        txndata.append(sharedData.data)


def module_run(txn: Txn):
    """
    Module Runnable Entry, be called when this module be picked up in current
     workflow

    @param txn Transaction context

    @return - True if no errors
            - False if error occurred
    """

    logger.debug("py module run")

    # Increase counters
    mod_metrics = metrics.module()
    mod_metrics.request.inc(1)
    return True
